File: src/lib/sabflow/actions/supabase.ts

'use server';

async function supabaseFetch(projectUrl: string, anonKey: string, method: string, path: string, body?: any, headers?: Record<string, string>, logger?: any) {
    logger?.log(`[Supabase] ${method} ${path}`);
    const options: RequestInit = {
        method,
        headers: {
            apikey: anonKey,
            Authorization: `Bearer ${anonKey}`,
            'Content-Type': 'application/json',
            Accept: 'application/json',
            Prefer: 'return=representation',
            ...headers,
        },
    };
    if (body !== undefined) options.body = JSON.stringify(body);
    const base = projectUrl.replace(/\/$/, '');
    const res = await fetch(`${base}${path}`, options);
    if (res.status === 204) return [];
    const data = await res.json();
    if (!res.ok) {
        throw new Error(data?.message || data?.hint || `Supabase API error: ${res.status}`);
    }
    return data;
}

export async function executeSupabaseAction(actionName: string, inputs: any, user: any, logger: any) {
    try {
        const projectUrl = String(inputs.projectUrl ?? '').trim();
        const anonKey = String(inputs.anonKey ?? '').trim();
        if (!projectUrl || !anonKey) throw new Error('projectUrl and anonKey are required.');
        const sb = (method: string, path: string, body?: any, headers?: Record<string, string>) =>
            supabaseFetch(projectUrl, anonKey, method, path, body, headers, logger);

        switch (actionName) {
            case 'select': {
                const table = String(inputs.table ?? '').trim();
                const columns = String(inputs.columns ?? '*').trim();
                const filter = String(inputs.filter ?? '').trim();
                const limit = inputs.limit ? Number(inputs.limit) : undefined;
                const orderBy = String(inputs.orderBy ?? '').trim();
                if (!table) throw new Error('table is required.');
                let path = `/rest/v1/${table}?select=${encodeURIComponent(columns)}`;
                if (filter) path += `&${filter}`;
                if (orderBy) path += `&order=${encodeURIComponent(orderBy)}`;
                if (limit) path += `&limit=${limit}`;
                const data = await sb('GET', path);
                return { output: { rows: Array.isArray(data) ? data : [data], count: Array.isArray(data) ? data.length : 1 } };
            }

            case 'insert': {
                const table = String(inputs.table ?? '').trim();
                const data_input = inputs.data;
                if (!table || !data_input) throw new Error('table and data are required.');
                const dataObj = typeof data_input === 'string' ? JSON.parse(data_input) : data_input;
                const result = await sb('POST', `/rest/v1/${table}`, dataObj);
                return { output: { inserted: Array.isArray(result) ? result : [result], count: String(Array.isArray(result) ? result.length : 1) } };
            }

            case 'update': {
                const table = String(inputs.table ?? '').trim();
                const data_input = inputs.data;
                const filter = String(inputs.filter ?? '').trim();
                if (!table || !data_input || !filter) throw new Error('table, data, and filter are required.');
                const dataObj = typeof data_input === 'string' ? JSON.parse(data_input) : data_input;
                const result = await sb('PATCH', `/rest/v1/${table}?${filter}`, dataObj);
                return { output: { updated: Array.isArray(result) ? result : [], count: String(Array.isArray(result) ? result.length : 0) } };
            }

            case 'upsert': {
                const table = String(inputs.table ?? '').trim();
                const data_input = inputs.data;
                if (!table || !data_input) throw new Error('table and data are required.');
                const dataObj = typeof data_input === 'string' ? JSON.parse(data_input) : data_input;
                const result = await sb('POST', `/rest/v1/${table}`, dataObj, { Prefer: 'return=representation,resolution=merge-duplicates' });
                return { output: { upserted: Array.isArray(result) ? result : [result] } };
            }

            case 'delete': {
                const table = String(inputs.table ?? '').trim();
                const filter = String(inputs.filter ?? '').trim();
                if (!table || !filter) throw new Error('table and filter are required.');
                const result = await sb('DELETE', `/rest/v1/${table}?${filter}`);
                return { output: { deleted: Array.isArray(result) ? result : [], count: String(Array.isArray(result) ? result.length : 0) } };
            }

            case 'rpc': {
                const functionName = String(inputs.functionName ?? '').trim();
                const params = inputs.params;
                if (!functionName) throw new Error('functionName is required.');
                const paramsObj = params ? (typeof params === 'string' ? JSON.parse(params) : params) : {};
                const result = await sb('POST', `/rest/v1/rpc/${functionName}`, paramsObj);
                return { output: { result } };
            }

            case 'signUpUser': {
                const email = String(inputs.email ?? '').trim();
                const password = String(inputs.password ?? '').trim();
                if (!email || !password) throw new Error('email and password are required.');
                const data = await sb('POST', '/auth/v1/signup', { email, password });
                return { output: { userId: data.user?.id ?? '', email: data.user?.email ?? '', accessToken: data.access_token ?? '' } };
            }

            case 'signInUser': {
                const email = String(inputs.email ?? '').trim();
                const password = String(inputs.password ?? '').trim();
                if (!email || !password) throw new Error('email and password are required.');
                const data = await sb('POST', '/auth/v1/token?grant_type=password', { email, password });
                return { output: { userId: data.user?.id ?? '', accessToken: data.access_token ?? '', refreshToken: data.refresh_token ?? '' } };
            }

            case 'getUser': {
                const accessToken = String(inputs.accessToken ?? '').trim();
                if (!accessToken) throw new Error('accessToken is required.');
                const res = await fetch(`${projectUrl.replace(/\/$/, '')}/auth/v1/user`, {
                    headers: { apikey: anonKey, Authorization: `Bearer ${accessToken}` },
                });
                const data = await res.json();
                return { output: { id: data.id ?? '', email: data.email ?? '', role: data.role ?? '' } };
            }

            case 'uploadFile': {
                const bucket = String(inputs.bucket ?? '').trim();
                const filePath = String(inputs.filePath ?? '').trim();
                const fileUrl = String(inputs.fileUrl ?? '').trim();
                if (!bucket || !filePath || !fileUrl) throw new Error('bucket, filePath, and fileUrl are required.');
                const fileRes = await fetch(fileUrl);
                const fileBlob = await fileRes.blob();
                const formData = new FormData();
                formData.append('', fileBlob, filePath.split('/').pop());
                const uploadRes = await fetch(`${projectUrl.replace(/\/$/, '')}/storage/v1/object/${bucket}/${filePath}`, {
                    method: 'POST',
                    headers: { apikey: anonKey, Authorization: `Bearer ${anonKey}` },
                    body: formData,
                });
                const uploadData = await uploadRes.json();
                if (!uploadRes.ok) throw new Error(uploadData?.message || 'Upload failed');
                return { output: { key: uploadData.Key ?? filePath, publicUrl: `${projectUrl}/storage/v1/object/public/${bucket}/${filePath}` } };
            }

            default:
                return { error: `Supabase action "${actionName}" is not implemented.` };
        }
    } catch (e: any) {
        return { error: e.message || 'Supabase action failed.' };
    }
}

File: rust/crates/sabflow-nodes/src/nodes/supabase.rs
//! Supabase node.
//!
//! Talks to a Supabase project over PostgREST (`{projectUrl}/rest/v1/{table}`).
//! Authenticates with either the service-role key (server-side, bypasses RLS)
//! or the anon key (subject to RLS). The chosen key is sent as both
//! `apikey: <key>` and `Authorization: Bearer <key>` headers.
//!
//! Operations:
//!   - `select` — GET   `/rest/v1/{table}?select={columns}&{filter}&limit={limit}`
//!   - `insert` — POST  `/rest/v1/{table}`  (body: JSON array or object)
//!   - `update` — PATCH `/rest/v1/{table}?{filter}`
//!   - `delete` — DELETE `/rest/v1/{table}?{filter}`
//!   - `upsert` — POST  `/rest/v1/{table}`  with `Prefer: resolution=merge-duplicates`
//!
//! Filters are passed through as raw PostgREST query strings, e.g.
//! `email=eq.user@example.com&status=in.(active,pending)`.

use async_trait::async_trait;
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use serde_json::{Value, json};

use crate::{
    context::{ExecutionContext, NodeInput, NodeOutput},
    descriptor::{
        CredentialBinding, NodeCategory, NodeDescriptor, NodeProperty, NodePropertyOption,
        NodePropertyType,
    },
    error::{NodeError, NodeResult},
    node::Node,
};

pub struct SupabaseNode;

#[async_trait]
impl Node for SupabaseNode {
    fn descriptor(&self) -> NodeDescriptor {
        NodeDescriptor::new(
            "supabase",
            "Supabase",
            "Supabase database operations via PostgREST",
            NodeCategory::Database,
        )
        .icon("database")
        .color("#3ECF8E")
        .credentials(vec![CredentialBinding {
            name: "supabaseApi".into(),
            display_name: "Supabase Project".into(),
            required: true,
        }])
        .properties(vec![
            NodeProperty::new("credentialId", "Credential", NodePropertyType::Credential)
                .required(),
            NodeProperty::new("table", "Table", NodePropertyType::String)
                .placeholder("users")
                .required(),
            NodeProperty::new("operation", "Operation", NodePropertyType::Options)
                .options(vec![
                    NodePropertyOption {
                        name: "Select".into(),
                        value: json!("select"),
                        description: Some("Fetch rows".into()),
                    },
                    NodePropertyOption {
                        name: "Insert".into(),
                        value: json!("insert"),
                        description: Some("Create new rows".into()),
                    },
                    NodePropertyOption {
                        name: "Update".into(),
                        value: json!("update"),
                        description: Some("Update rows matching filter".into()),
                    },
                    NodePropertyOption {
                        name: "Delete".into(),
                        value: json!("delete"),
                        description: Some("Delete rows matching filter".into()),
                    },
                    NodePropertyOption {
                        name: "Upsert".into(),
                        value: json!("upsert"),
                        description: Some("Insert or update on conflict".into()),
                    },
                ])
                .default(json!("select"))
                .required(),
            NodeProperty::new("filter", "Filter", NodePropertyType::String)
                .placeholder("email=eq.user@example.com")
                .description(
                    "PostgREST filter — e.g. id=eq.5, status=in.(active,pending). \
                     Multiple filters joined with `&`.",
                )
                .show_when("operation", &["select", "update", "delete"]),
            NodeProperty::new("data", "Data", NodePropertyType::Json)
                .description("Row data — JSON object or array of objects")
                .default(json!({}))
                .show_when("operation", &["insert", "update", "upsert"]),
            NodeProperty::new("columns", "Columns", NodePropertyType::String)
                .default(json!("*"))
                .placeholder("id,name,email")
                .description("Comma-separated columns for SELECT")
                .show_when("operation", &["select"]),
            NodeProperty::new("limit", "Limit", NodePropertyType::Number)
                .default(json!(100))
                .show_when("operation", &["select"]),
        ])
    }

    async fn execute(
        &self,
        ctx: &mut ExecutionContext,
        _input: NodeInput,
        params: &Value,
    ) -> NodeResult<NodeOutput> {
        let cred_id = ctx.param_str(params, "credentialId")?;
        let cred = ctx.credential(&cred_id)?;
        let project_url = cred
            .data
            .get("projectUrl")
            .map(|s| s.trim_end_matches('/').to_string())
            .ok_or_else(|| NodeError::MissingParameter("projectUrl".into()))?;
        // Either service-role or anon key works; prefer service-role.
        let api_key = cred
            .data
            .get("serviceRoleKey")
            .or_else(|| cred.data.get("anonKey"))
            .cloned()
            .ok_or_else(|| {
                NodeError::MissingParameter("serviceRoleKey or anonKey".into())
            })?;

        let table = ctx.substitute(&ctx.param_str(params, "table")?);
        if table.is_empty() {
            return Err(NodeError::MissingParameter("table".into()));
        }
        let operation = ctx.param_str(params, "operation")?;

        let base = format!(
            "{base}/rest/v1/{table}",
            base = project_url,
            table = urlencoding::encode(&table),
        );

        let mut headers = HeaderMap::new();
        let key_val = HeaderValue::from_str(&api_key)
            .map_err(|e| NodeError::AuthError(format!("invalid api key: {e}")))?;
        headers.insert(HeaderName::from_static("apikey"), key_val.clone());
        let bearer = HeaderValue::from_str(&format!("Bearer {api_key}"))
            .map_err(|e| NodeError::AuthError(format!("invalid bearer: {e}")))?;
        headers.insert(reqwest::header::AUTHORIZATION, bearer);

        match operation.as_str() {
            "select" => {
                let cols = ctx
                    .param_str_opt(params, "columns")
                    .map(|s| ctx.substitute(s.trim()))
                    .filter(|s| !s.is_empty())
                    .unwrap_or_else(|| "*".to_string());
                let limit = ctx.param_f64(params, "limit").unwrap_or(100.0) as i64;
                let filter = ctx
                    .param_str_opt(params, "filter")
                    .map(|s| ctx.substitute(s.trim()))
                    .filter(|s| !s.is_empty());

                let mut url = format!(
                    "{base}?select={cols}&limit={limit}",
                    base = base,
                    cols = urlencoding::encode(&cols),
                    limit = limit,
                );
                if let Some(f) = filter {
                    // Filter is a raw PostgREST string. We append as-is so
                    // multi-filter expressions stay intact.
                    url.push('&');
                    url.push_str(&f);
                }

                let res = ctx
                    .http
                    .get(&url)
                    .headers(headers)
                    .send()
                    .await?;
                read_json(res).await
            }
            "insert" => {
                let body = read_json_value(ctx, params, "data")?;
                let mut req_headers = headers;
                req_headers.insert(
                    HeaderName::from_static("prefer"),
                    HeaderValue::from_static("return=representation"),
                );
                req_headers.insert(
                    reqwest::header::CONTENT_TYPE,
                    HeaderValue::from_static("application/json"),
                );
                let res = ctx
                    .http
                    .post(&base)
                    .headers(req_headers)
                    .json(&body)
                    .send()
                    .await?;
                read_json(res).await
            }
            "update" => {
                let filter = ctx
                    .param_str_opt(params, "filter")
                    .map(|s| ctx.substitute(s.trim()))
                    .filter(|s| !s.is_empty())
                    .ok_or_else(|| NodeError::MissingParameter("filter".into()))?;
                let body = read_json_value(ctx, params, "data")?;
                let url = format!("{base}?{filter}");
                let mut req_headers = headers;
                req_headers.insert(
                    HeaderName::from_static("prefer"),
                    HeaderValue::from_static("return=representation"),
                );
                req_headers.insert(
                    reqwest::header::CONTENT_TYPE,
                    HeaderValue::from_static("application/json"),
                );
                let res = ctx
                    .http
                    .patch(&url)
                    .headers(req_headers)
                    .json(&body)
                    .send()
                    .await?;
                read_json(res).await
            }
            "delete" => {
                let filter = ctx
                    .param_str_opt(params, "filter")
                    .map(|s| ctx.substitute(s.trim()))
                    .filter(|s| !s.is_empty())
                    .ok_or_else(|| NodeError::MissingParameter("filter".into()))?;
                let url = format!("{base}?{filter}");
                let mut req_headers = headers;
                req_headers.insert(
                    HeaderName::from_static("prefer"),
                    HeaderValue::from_static("return=representation"),
                );
                let res = ctx
                    .http
                    .delete(&url)
                    .headers(req_headers)
                    .send()
                    .await?;
                read_json(res).await
            }
            "upsert" => {
                let body = read_json_value(ctx, params, "data")?;
                let mut req_headers = headers;
                req_headers.insert(
                    HeaderName::from_static("prefer"),
                    HeaderValue::from_static(
                        "resolution=merge-duplicates,return=representation",
                    ),
                );
                req_headers.insert(
                    reqwest::header::CONTENT_TYPE,
                    HeaderValue::from_static("application/json"),
                );
                let res = ctx
                    .http
                    .post(&base)
                    .headers(req_headers)
                    .json(&body)
                    .send()
                    .await?;
                read_json(res).await
            }
            other => Err(NodeError::InvalidParameter {
                name: "operation".into(),
                reason: format!("unknown operation: {other}"),
            }),
        }
    }
}

/// Read a JSON-typed parameter. Accepts a JSON object/array directly or a
/// JSON-encoded string (with `{{var}}` substitution applied to the string form).
fn read_json_value(ctx: &ExecutionContext, params: &Value, key: &str) -> NodeResult<Value> {
    let raw = params.get(key).cloned().unwrap_or(Value::Null);
    match raw {
        Value::String(s) => {
            let sub = ctx.substitute(&s);
            let trimmed = sub.trim();
            if trimmed.is_empty() {
                Ok(Value::Object(serde_json::Map::new()))
            } else {
                serde_json::from_str::<Value>(trimmed).map_err(|e| {
                    NodeError::InvalidParameter {
                        name: key.into(),
                        reason: format!("not valid JSON: {e}"),
                    }
                })
            }
        }
        Value::Null => Ok(Value::Object(serde_json::Map::new())),
        other => Ok(other),
    }
}

/// Parse a Supabase REST response. Non-2xx responses surface as
/// `UpstreamError`. 204 (no body, e.g. DELETE without Prefer) returns an
/// empty branch. The success body is wrapped so that an array of rows becomes
/// multiple items and a single object becomes one item.
async fn read_json(res: reqwest::Response) -> NodeResult<NodeOutput> {
    let status = res.status();
    if status.as_u16() == 204 {
        return Ok(NodeOutput::single(vec![]));
    }
    let text = res.text().await.unwrap_or_default();
    if !status.is_success() {
        return Err(NodeError::UpstreamError {
            status: status.as_u16(),
            body: text,
        });
    }
    if text.trim().is_empty() {
        return Ok(NodeOutput::single(vec![]));
    }
    let parsed: Value = serde_json::from_str(&text).map_err(|e| {
        NodeError::SerializationError(format!("supabase response: {e}; body: {text}"))
    })?;
    let items = match parsed {
        Value::Array(arr) => arr,
        other => vec![other],
    };
    Ok(NodeOutput::single(items))
}
